### ISSUE TITLE
scRNA Discovery step2 aborts after `state_discovery_scRNA_distances` 

### ISSUE THREAD

**Issue by NoahHenrikKleinschmidt** [digitalcytometry/ecotyper#41]
Hello there, 

This is a follow-up to my earlier issue that I self-diagnosed and is closed now. 

So, after a lot of diagnosing I found that step 2 (which has three sub-steps) terminates after the first sub-step (calling `state_discovery_scRNA_distances` )  although without any error message that I can discern from the output. 
However, after commenting out the first sub-step and re-running the pipeline finishes successfully. 

The code in question is:

```{r}
for(cell_type in cell_types)
	{
		filter_genes = (fractions == "Cell_type_specific_genes")
		PushToJobQueue(paste("Rscript state_discovery_scRNA_distances.R", discovery, fractions, cell_type, filter_genes, scale_column))	
	}	
	log_info( "(sub1) Running jobs..." )
	RunJobQueue()                                   # <- this line "breaks" without any recognizable error message
        log_info( "(sub1) Finished jobs..." )    # and thus this line is never logged
```

However, after an unsuccessful run of step2 that broke at the marked line, commenting the code out allows all subsequent steps to seemingly finish successfully...

```{r}
for(cell_type in cell_types)
	# {
	# 	filter_genes = (fractions == "Cell_type_specific_genes")
	# 	log_info( "filter_genes = ")
	# 	print( filter_genes )
	# 	PushToJobQueue(paste("Rscript state_discovery_scRNA_distances.R", discovery, fractions, cell_type, filter_genes, scale_column))	
	# }	
	# log_info( "Running jobs..." )
	# RunJobQueue()                                # <- now nothing happens here of course
        log_info( "(sub1) Finished jobs..." )    #  this line is logged and all subsequent steps seem to work
```

Would you happen to know about this behavior and how I can avoid it? 

Thanks and best,
Noah

**Comment by BALuca** [comment]
Hi Noah, 

Thank you for reporting the issue. The line you are commenting out is actually important to make sure that all parallel jobs launched inside the loop above that line finish, before moving on to the next step. Judging by the errors in the first ticket you created and this one, I would say that there is a formatting issue with the input data. One way you could debug it yourself is to make sure that if you run the following R code you don't get any error:

```
tmp = read.delim("your expression matrix")
tmp = tmp + 1
```
If problem persists, would you be able to send us at ecotyper.stanford@gmail.com a small sample of your data in which the issue is reproducible, to debug the issue?

Best,
The EcoTyper team

**Comment by NoahHenrikKleinschmidt** [comment]
Thanks a lot for the reply. 
I have tried the code and indeed, I got an error:

```
> mydata = read.delim(f)
> mydata1 = mydata + 1
Error in FUN(left, right) :
  non-numeric argument for binary operator
```

As it turned out the first column within my datafile are the gene identifiers, which of course are non-numeric but `character`. 
I was able to fix the issue by using 

```
> names = make.names( mydata[,1], unique = TRUE )
> rownames( mydata ) = names
> mydata = mydata[,-1]
>
> mydata1 = mydata + 1 
```
Now the line works just fine.

Since the first column is supposed to be gene identifiers I was next wondering why I got this error in the first place. I was able to diagnose the issue. The first column was actually named. 

![Bildschirmfoto 2022-08-06 um 17 10 48](https://user-images.githubusercontent.com/89252165/183255125-bfad14bf-b925-43f3-90d2-638c03b58928.png)

I *removed this header* from the datafile and repeated the same procedure.  Now the code snippet worked without issue. 

![Bildschirmfoto 2022-08-06 um 17 11 43](https://user-images.githubusercontent.com/89252165/183255130-819d4fef-a59e-4051-b7c5-f949199763d9.png)


As I recall from Tutorial 5, the example input data also seems to have had a named first column, however, although the name as `Gene` rather than `name`. Does this mean EcoTyper is hard-coded to recognize `Gene` as a valid header for the gene-identifier column?

I have re-started EcoTyper again with this file and now received the following warning:

```
Warning message:
In fread(input_mat, sep = "\t", nrows = 5) :
  Detected 11265 column names but the data has 11266 columns (i.e. invalid file). Added 1 extra default column name for the first column which is guessed to be row names or an index. Use setnames() afterwards if this guess is not correct, or fix the file write command that created the file to create a valid file.
```

The pipeline seems to run fine but again crash at the same `RunJobQueue()` call as before...

Also, I noticed that for some reason one of the cell types is just called `TRUE`? I searched a bit more through the files, and while there are no `"TRUE_cell_type_specific_genes.txt"` generated there are `"NA_cell_type_specific_genes.txt"`. I found that there are some entries within the annotation file that do not have a cell type specified (i.e. are blank in the CellType column). Could it be possible that these samples are the cause for the issue?
